### PR TITLE
overlord/ifacestate: introduce HotplugKey type use short key in change summaries

### PR DESCRIPTION
**Pull request by stolowski**
Introduce HotplugKey type based on string and use it everywhere in hotplug code instead of plain string. Provide ShortString method (similiar to recently landed ShortString method of HotplugDeviceInfo) and use it for task and change summaries; debug messages will still use full hotplug key string. Also fixed a few inconsistencies with "%s" vs "%q" in the messages.

Per one of the review comments suggesting use of unicode … for "...", I've also updated HotplugDeviceKey the same way.

### CODE REVIEW

File: interfaces/repo.go [hunk 1]
```diff
@@ -817,7 +817,7 @@ func (r *Repository) UpdateHotplugSlotAttrs(ifaceName, hotplugKey string, static
 	}
 
 	for _, slotInfo := range r.slots[snapName] {
-		if slotInfo.Interface == ifaceName && slotInfo.HotplugKey == hotplugKey {
+		if slotInfo.Interface == ifaceName && (slotInfo.HotplugKey == hotplugKey) {
```
Comment (zyga) [inline interfaces/repo.go]: Is this change relevant?

File: overlord/ifacestate/hotplug_test.go [hunk 2]
```diff
@@ -1120,14 +1120,14 @@ func (s *hotplugSuite) TestRemoveDeviceTasks(c *C) {
...
-	c.Assert(key, Equals, "key")
+	c.Assert(key, DeepEquals, snap.HotplugKey("key"))
 
 	task2 := tss.Tasks()[1]
 	c.Assert(task2.Kind(), Equals, "hotplug-remove-slot")
 	iface, key, err = ifacestate.GetHotplugAttrs(task2)
 	c.Assert(err, IsNil)
 	c.Assert(iface, Equals, "interface")
-	c.Assert(key, Equals, "key")
+	c.Assert(key, DeepEquals, snap.HotplugKey("key"))
```
Comment (zyga) [inline overlord/ifacestate/hotplug_test.go]: Why does this now require deep equals? Is this due to the fact that the string is wrapped in a new type?

File: overlord/ifacestate/ifacestate_test.go [hunk 3]
```diff
@@ -5707,7 +5707,7 @@ func (s *interfaceManagerSuite) TestHotplugAddGoneSlot(c *C) {
 	slot := repo.Slot("core", "hotplugslot-old-name")
 	c.Assert(slot, NotNil)
 	c.Check(slot.Attrs, DeepEquals, map[string]interface{}{"foo": "bar"})
-	c.Check(slot.HotplugKey, Equals, "1234")
+	c.Check(slot.HotplugKey, DeepEquals, snap.HotplugKey("1234"))
```
Comment (zyga) [inline overlord/ifacestate/ifacestate_test.go]: Why is this the case that just above this changed fragment we can use `Equals` and here the test was updated to `DeepEquals`?

File: snap/hotplug_key.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package snap
+
+// HotplugKey is a string key of a hotplugged device
+type HotplugKey string
+
+// ShortString returns a truncated string representation of the hotplug key
+func (h HotplugKey) ShortString() string {
+	str := string(h)
```
Comment (zyga) [inline snap/hotplug_key.go]: I have not looked around the tree but perhaps this should be a common helper that is in `strutil`?

File: snap/hotplug_key.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	// this is normally never the case, but makes testing easier
+	if len(str) < 12 {
```
Comment (zyga) [inline snap/hotplug_key.go]: I don't doubt this value was picked carefully, could we have the reasoning of the value added as a code comment please?

File: snap/hotplug_key.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+		return str
+	}
+	return str[0:12] + "..."
```
Comment (zyga) [inline snap/hotplug_key.go]: Should this be using the unicode horizontal ellipsis character U+2026? CC @chipaca

File: interfaces/repo.go (hunk 1 above)
Comment (stolowski) [reply interfaces/repo.go]: Ah, not really, restored, thanks for spotting!

File: overlord/ifacestate/hotplug_test.go (hunk 2 above)
Comment (stolowski) [reply overlord/ifacestate/hotplug_test.go]: It doesn't require DeepEquals, I just felt it's better for custom types (in particular if they ever grow up, probably unlikely here). I can revert back to Equals if you think that's confusing?

File: overlord/ifacestate/ifacestate_test.go (hunk 3 above)
Comment (stolowski) [reply overlord/ifacestate/ifacestate_test.go]: Not really needed, but see earlier explanation.

File: snap/hotplug_key.go (hunk 5 above)
Comment (stolowski) [reply snap/hotplug_key.go]: I'm not sure it deserves a helper atm, we have two very specific use cases (HotplugKey and HotplugDeviceInfo - the latter truncates only vendor/model as part of the final string).

File: snap/hotplug_key.go (hunk 6 above)
Comment (stolowski) [reply snap/hotplug_key.go]: Done (also updated it for HotplugDeviceInfo).

File: interfaces/hotplug/deviceinfo.go [hunk 7]
```diff
@@ -120,7 +120,7 @@ func (h *HotplugDeviceInfo) str(maxModelOrVendorLen int) string {
 
 	modelOrVendor := h.firstAttrValueOf("ID_MODEL_FROM_DATABASE", "ID_MODEL", "ID_MODEL_ID", "ID_VENDOR_FROM_DATABASE", "ID_VENDOR", "ID_VENDOR_ID")
 	if len(modelOrVendor) > maxModelOrVendorLen {
-		modelOrVendor = modelOrVendor[0:maxModelOrVendorLen] + "..."
+		modelOrVendor = modelOrVendor[0:maxModelOrVendorLen] + "…"
```
Comment (bboozzoo) [inline interfaces/hotplug/deviceinfo.go]: This should display correctly even in non-utf capable terminal, right?

File: snap/hotplug_key.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package snap
+
+import (
+	"fmt"
+)
+
+// HotplugKey is a string key of a hotplugged device
+type HotplugKey string
```
Comment (bboozzoo) [inline snap/hotplug_key.go]: Likie it way more than the bare string that carries no information about the purpose or format.

File: interfaces/hotplug/deviceinfo.go (hunk 7 above)
Comment (stolowski) [reply interfaces/hotplug/deviceinfo.go]: No really, it will appear as a rhombus, but as discussed on IRC that's fine.

**Pull request merged**